[quant] Modify APoT nonuniform quantization workflow

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79845
* __->__ #80075
* #80303

### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

asl3 committed:
[quant] Modify APoT Observer class max_val calculation

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer class max_val calculation"




[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks like we need to change the forward to record both max_val and min_val as well

asl3 commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -23,16 +23,19 @@ class APoTObserver(ObserverBase):
             dtype=torch.quint8) -> None:
         super().__init__(dtype)
-        self.max_val = max_val
+        self.max_val = 1.0
```
agreed

asl3 committed:
Update on "[quant] Modify APoT Observer class max_val calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer class alpha calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes. The PR also modifies all occurrences of calls to `max_val` and `APoTObserver` constructor accordingly

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer class alpha calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes. The PR also modifies all occurrences of calls to `max_val` and `APoTObserver` constructor accordingly

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer class alpha calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes. The PR also modifies all occurrences of calls to `max_val` and `APoTObserver` constructor accordingly

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer class alpha calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes. The PR also modifies all occurrences of calls to `max_val` and `APoTObserver` constructor accordingly

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -18,6 +18,7 @@ def __init__(
         self,
         b,
         k,
+        min_val,
```
the quantizer should take quantization parameters as arguments, e.g. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/quantized/Quantizer.cpp#L70-L73

asl3 replied:
@dzdang that makes sense to me, i'll do that!

dzdang replied:
it doesn't seem this was addressed in the latest version

asl3 replied:
@dzdang my quantizer `init` function is only taking in observer now -- is that what you meant above?

asl3 replied:
@dzdang i interpreted the above as just passing in observer to quantizer, but i see you mean that externally observer should compute quantization parameters and then pass those into constructor for quantizer. will fix!

asl3 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -18,6 +18,7 @@ def __init__(
+        min_val,
         max_val,
         signed,
             dtype=torch.quint8) -> None:
```
@dzdang agreed, i will change to torch.uint8

asl3 committed:
Update on "[quant] Modify APoT Observer class alpha calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes. The PR also modifies all occurrences of calls to `max_val` and `APoTObserver` constructor accordingly

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer class alpha calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes. The PR also modifies all occurrences of calls to `max_val` and `APoTObserver` constructor accordingly

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer class alpha calculation"


### Summary:
This PR modifies the way the `max_val` attribute is calculated by an APoT Observer. max_val is no longer passed into the APoT Observer constructor, but rather calculated using the APoT `forward` method. The APoT `calculate_qparams` method also accepts an optional max_val parameter that can override the internal `max_val` attribute for testing purposes. The PR also modifies all occurrences of calls to `max_val` and `APoTObserver` constructor accordingly

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer, Quantizer, and Tensor class design"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified to remove explicit dependency of quantization parameters, and instead take in an APoT Observer as input. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer, Quantizer, and Tensor class design"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified to remove explicit dependency of quantization parameters, and instead take in an APoT Observer as input. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer, Quantizer, and Tensor class design"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified to remove explicit dependency of quantization parameters, and instead take in an APoT Observer as input. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer, Quantizer, and Tensor class design"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified to remove explicit dependency of quantization parameters, and instead take in an APoT Observer as input. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer, Quantizer, and Tensor class design"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified to remove explicit dependency of quantization parameters, and instead take in an APoT Observer as input. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer, Quantizer, and Tensor class design"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified to remove explicit dependency of quantization parameters, and instead take in an APoT Observer as input. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT Observer, Quantizer, and Tensor class design"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified to remove explicit dependency of quantization parameters, and instead take in an APoT Observer as input. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -95,8 +96,8 @@ def test_calculate_qparams_3terms(self):
-        obs = APoTObserver(max_val=1.0, b=4, k=2)
-        obs_result = obs.calculate_qparams(signed=True)
+        obs = APoTObserver(b=4, k=2)
+        obs_result = obs.calculate_qparams(signed=True, min_val=torch.tensor([0]), max_val=torch.tensor([1]))
```
nit: maybe we can just unpack the individual qparams here so that it's clearer

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -47,28 +33,41 @@ def __init__(
+
+r""" Global method to create quantizer and call quantizer quantize_APoT
+"""
+def quantize_APoT(tensor2quantize: Tensor, qparams: Tuple):
```
nit: or we can unpack the qparams args here as well, like
```
quantize_APoT(tensor2quantize: Tensor, alpha: ..., gamma: ..., quantization_levels: ..., ...)
```

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -47,28 +33,41 @@ def __init__(
+
+r""" Global method to create quantizer and call quantizer dequantize_APoT
+"""
+def dequantize_APoT(float2apot, qparams: Tuple):
+    quantizer = APoTQuantizer(alpha=qparams[0], gamma=qparams[1], quantization_levels=qparams[2], level_indices=qparams[3])
+    return quantizer.dequantize_APoT(float2apot)
```
`float2apot` --> `apot_tensor`?

jerryzh168 replied:
also it can be:
```
def deqauntize_APoT(apot_tensor):
        return apot_tensor.quantizer.dequantize(apot_tensor)
```
please add type annotations as well

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -47,28 +33,41 @@ def __init__(
     def quantize_APoT(self, tensor2quantize: Tensor):
```
can you add type annotation for output as well?

asl3 replied:
@jerryzh168 i didn't include type annotations for the return value (TensorAPoT) because when I do I get a circular import error (quantizer imports APoT_Tensor and APoT_Tensor imports quantizer). is it necessary to specify return value type? if so i can try to figure out another solution

asl3 replied:
note: after discussion decided to just leave comment explaining APoT Tensor type

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/experimental/apot_utils.py:
```diff
@@ -48,3 +48,11 @@ def float_to_reduced_precision(x, levels, indices):
+
+r"""Clips value x based on alpha qparam
+"""
+def clip(x, alpha):
```
can just use torch.clamp

HDCharles replied:
also, you should either be clipping positive and negative, or not be clipping at all.

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,20 +2,21 @@
 
         with self.assertRaises(AssertionError):
-            obs_result = obs.calculate_qparams(signed=False)
+            obs_result = obs.calculate_qparams(signed=False, min_val=torch.tensor([0]), max_val=torch.tensor([0]))
```
seems Jerry's comment about unpacking individual qparams was not addressed. 
a bit clear (so that a reader doesn't have to dig into what calculate_qparams is returning) to just do
`alpha, gamma, quantized_levels, level-indices = obs.calculate_qparams(...)`
here

asl3 replied:
oh yes changed it in other tests but missed this file, will update!

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Modify APoT nonuniform quantization workflow"


### Summary:
This PR updates the design of APoT Observer, Quantizer, and Tensor to be more consistent with their uniform counterparts in the PyTorch framework. APoT Observer now calculates alpha as the max between the absolute values of the max and min values in the input tensor. APoT Quantizer is modified so its instance methods quantize_APoT and dequantize_APoT are called by their global method counterparts. APoT Tensor is modified to account for the new method definition of the `quantize_APoT` from APoT Quantizer.

### Test Plan:
Run APoT Observer class unit tests with: `python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py`
Run APoT Quantize class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`
Run APoT Tensor class unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 commented:
@pytorchbot merge -g
